### ISSUE TITLE
The data returned by the get_any method in variable_pool.py has not deeply restored the Varibale type, causing an error in string concatenation in llm_node.

### ISSUE THREAD

**Issue by fictionking** [langgenius/dify#6662]
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.6.11

### Cloud or Self Hosted

Self Hosted (Source)

### Steps to reproduce

The data returned by the get_any method in variable_pool.py has not deeply restored the Varibale type, causing an error in string concatenation in llm_node.

### Suggest:
add deep restore method in api/core/app/segments/factory.py
```
def retore_variable(var: Any) -> Any:
    if isinstance(var, StringVariable):
        return var.value
    elif isinstance(var, IntegerVariable):
        return var.value
    elif isinstance(var, FloatVariable):
        return var.value
    elif isinstance(var, ObjectVariable):
        return {k: retore_variable(v) for k, v in var.value.items()}
    elif isinstance(var, ArrayVariable):
        return [retore_variable(v) for v in var.value]
    elif isinstance(var, FileVariable):
        return var.value
    return var
```

modify api/core/workflow/entities/variable_pool.py
``` 
def get_any(self, selector: Sequence[str], /) -> Any | None:

        .....

        if value is None:
            return value
        if isinstance(value, Variable):
            value=factory.retore_variable(value)
        return value
```

**Comment by laipz8200** [comment]
Could you please provide some specific examples? The `get_any` method restores each value recursively by invoking the Segment class's `to_object()` method.

We have included some tests for this method in `tests/unit_tests/app/test_variables.py`.

**Comment by fictionking** [comment]
@laipz8200 

I fund in api/core/app/segments/factory.py has some TODO below:
```
def build_anonymous_variable(value: Any, /) -> Variable:
    if isinstance(value, str):
        return StringVariable(name='anonymous', value=value)
    if isinstance(value, int):
        return IntegerVariable(name='anonymous', value=value)
    if isinstance(value, float):
        return FloatVariable(name='anonymous', value=value)
    if isinstance(value, dict):
        # TODO: Limit the depth of the object
        obj = {k: build_anonymous_variable(v) for k, v in value.items()}
        return ObjectVariable(name='anonymous', value=obj)
    if isinstance(value, list):
        # TODO: Limit the depth of the array
        elements = [build_anonymous_variable(v) for v in value]
        return ArrayVariable(name='anonymous', value=elements)
    if isinstance(value, FileVar):
        return FileVariable(name='anonymous', value=value)
    raise ValueError(f'not supported value {value}')
```
completed this TODO,this issue will not happend!

**Comment by fictionking** [comment]
this is a simple sample data to test:
```json
[
{'metadata': {'_source': 'knowledge', 'position': 1}, 'title': 'aaa.pdf', 'content': 'some contents'},
{'metadata': {'_source': 'knowledge', 'position': 1}, 'title': 'aaa.pdf', 'content': 'some contents'}
]
```

**Comment by fictionking** [comment]
this is my workflow:
![屏幕截图 2024-07-25 171154](https://github.com/user-attachments/assets/2fa8ab4c-c3a7-4912-9938-5b7faaa84998)

**Comment by laipz8200** [comment]
@fictionking We've implemented recursively access in #6572, and the example data you gave above can pass the test like this:

```python3
def test():
    # It can restore the original value
    value = [
        {'metadata': {'_source': 'knowledge', 'position': 1}, 'title': 'aaa.pdf', 'content': 'some contents'},
        {'metadata': {'_source': 'knowledge', 'position': 1}, 'title': 'aaa.pdf', 'content': 'some contents'},
    ]
    var = factory.build_anonymous_variable(value)
    assert var.to_object() == value

    # Also for the variable pool
    pool = VariablePool(system_variables={}, user_inputs={}, environment_variables=[])
    selector = ['node_id', 'key_name']
    pool.add(selector, var)
    assert pool.get_any(selector) == value

```

You may need to update to the latest version and then try again.
